### ISSUE TITLE
Using this code with picture and pointcloud

### ISSUE THREAD

**Issue by zhengshunkai** [dylan-campbell/bpnpnet#7]
Hi! Thank you so much for your code.

I have a data set that consists of:

A set of 3d points (xyz values); 
A set of images
A camera matrix 

Is it possible to run your code with this data?
By the way,How to convert images to 2D points？Through visual features Such as FastCornerPoints,Harris,Orb,Sift,Surf ....?

Thanks!

**Comment by dylan-campbell** [comment]
Hi Shunkai, it works best if you use the same 2D detector as was used to generate the 3D points. For example, if you use the sift detector and feature extractor to generate the SfM point cloud, then use the sift detector to obtain the 2D points from your new images. If the 3D points come from lidar/rgbd rather than SfM, then it's best to use a detector that extracts geometrically-meaningful features near edges/corners in the image - pick your favourite detector. Even a Canny edge detector can work pretty well for obtaining useful 2D points!

**Comment by zhengshunkai** [comment]
> Hi Shunkai, it works best if you use the same 2D detector as was used to generate the 3D points. For example, if you use the sift detector and feature extractor to generate the SfM point cloud, then use the sift detector to obtain the 2D points from your new images. If the 3D points come from lidar/rgbd rather than SfM, then it's best to use a detector that extracts geometrically-meaningful features near edges/corners in the image - pick your favourite detector. Even a Canny edge detector can work pretty well for obtaining useful 2D points!

Thank you for your reply..I know how to get the 2D points.
Actualy I have a date set that consists of :
PointCloudMap (**global x y z values**);
A set of images; 
The init pose of images;

what I want to do is :）
Obtain a part of  PointCloudMap through the initial cam pose (**global Terror**) and  then project the map to the camera coordinate system to get the 3D points. Obtain the 2D points by Fast or Canny edge detector . Then use your great net to get the Optimized pose **Topt** via 2D-3D match.

**Comment by dylan-campbell** [comment]
Ok, yes that should work fine. It's not designed specifically for local optimisation like (it solves the global pose problem - ie without having an initial camera pose), so there are likely to be better solvers for your specific local optimisation problem, but it should still work decently.

**Comment by zhengshunkai** [comment]
Hello，I have some problems here:
1.I use the 3D points(xp,yp,zp) and 2D points maked by this 3D points （xp/zp,yp/zp）,The normalized coordinates of the image are as follows
![image](https://user-images.githubusercontent.com/9248030/192187079-b650de45-81c5-443b-b500-8e7b26fa0b6f.png)

2.Then use the  data61_2d3d_LcLp.pth.tar for test.
But the result is bad..The T error is very big (>3m)..

**Comment by dylan-campbell** [comment]
Hi, you won't be able to use the pre-trained weights for your data - it is very different to the training data. You'll need to train on the type of data you want to use. Note that the pre-processing used for the Data61/2D3D dataset was as follows (and might be helpful in your case as well) [copied from the paper appendix]:
> Each panoramic image was converted to a set of regular images with
> a 72◦ field-of-view to increase the challenge of the dataset. We used a semantic
> segmentation to extract, in 2D and 3D, static objects that do not lie on the
> ground plane, for example, buildings and trees, and then randomly downsample
> to 1000 points and pixels.

**Comment by zhengshunkai** [comment]
OK.I will train the model.
By the way,can you explain the W_gt format？The training must use the W_ gt?
Thank you for your reply :)

**Comment by dylan-campbell** [comment]
Yes, training needs W_gt - it's the 2D-3D correspondences, formatted as a list of indices (see dataset_utilities.py for more details).

**Comment by zhengshunkai** [comment]
I know it is a list of indices such as below,but I dont know what it means.How to express matching relationship in one-dimensional list？The first N values are 2d matches index and the end N values are 3d matches index or two adjacent values represent 2d-3d matched index?

1: array([362,  80, 874, 456, 213, 852, 949, 240, 685, 645, 481, 182, 326,
       914, 354, 945, 337, 100, 270, 402, 468,  40, 880, 508, 514, 355,
       287, 404,  55, 872, 666, 940, 415, 907,  44, 538, 343, 184, 450,
       228, 835, 913, 687,  19, 236, 315, 437, 920, 569, 365, 201, 703,
       541, 251, 705, 676, 368, 868, 266,  61, 364, 341, 603,   1, 420,
       557, 158, 744, 989, 987, 642, 578, 782,   2, 867, 710, 179, 206,
        99, 718, 300,  36, 563, 785, 711, 391, 458, 554, 609, 904, 936,
        37, 318, 749, 757, 719, 574, 492, 923, 751,  46,  11, 461, 755,
       985, 399, 524, 200, 929, 691, 202, 905, 927, 900, 390, 302, 306,
       552, 321, 671,  86,  95, 579, 441, 686, 503, 632, 457, 400, 976,
       667, 486, 248, 697, 803, 243, 303, 591, 848, 690, 698, 209, 890,
        43, 523, 194, 693,  38, 674, 947, 560, 594, 159, 922, 430, 804,
       380, 361, 309, 373, 857, 446, 934, 408, 153, 919, 440, 128, 314,
       198, 211, 286, 405, 752, 608,  34, 973, 511, 162, 290, 442, 729,
        63, 717, 599, 484, 369, 866, 336, 754, 532,  54,  96, 230, 564,
        73, 549, 571, 595, 298, 738,  35,  78, 347, 521, 103, 233, 901,
       467, 299, 800, 410,  71, 853, 897, 863, 428, 854, 670, 884, 968,
       891, 996,   9, 917, 994, 826, 995, 407, 975,  84, 633, 247, 487,
       692,  79, 836, 227, 109, 962, 261, 583, 423,  74, 653, 498, 547,
       489,  82, 921, 389, 443, 830, 469, 146, 255, 980, 781, 604, 783,
       433, 108, 742, 191, 464,  14, 546, 743,  91, 842, 728, 733, 127,
         8, 414, 387, 241, 797, 105, 359, 186, 371, 528, 246, 385, 439,
        83, 951,  70, 172, 824, 983, 655, 165, 518, 392, 135,   3, 592,
       739, 259, 155, 148,  72, 141, 379, 431, 421, 322, 834,  87, 831,
       898, 265,  85, 636, 963, 522, 861, 292, 485, 795, 559, 886, 460,
       212, 709, 253, 409, 277, 374, 931, 136, 245, 131, 869, 562,  49,
       827,  50, 462, 881, 526, 548, 224, 851, 229, 307, 779,  23, 689,
       607, 493,  24, 664, 473, 999, 736,  88, 551, 844, 611, 696, 908,
       533, 984, 606, 225, 232, 101, 157, 406, 829, 871, 915, 911, 534,
       293, 121,  93, 500, 954, 764, 688, 325, 115, 539, 769, 143, 242,
       722, 894, 452, 273, 756, 220, 478, 892, 942, 397, 348, 714, 568,
       381, 961, 957, 778, 651, 239, 643, 208, 274, 634,  92, 593, 888,
       993, 878, 648, 449, 419, 896, 882, 491, 207, 821, 205, 616, 346,
       133, 570, 828, 388, 174, 173, 810, 496, 187, 288, 543, 780, 699,
       125, 363, 625, 793, 669, 432, 175, 860, 502, 537, 338, 216, 590,
       142, 145, 150, 820, 340, 335, 497, 573, 114, 112, 575, 117, 475,
       683, 740, 775, 903, 864, 351, 342, 178,   4, 377, 221, 635, 938,
       614, 250, 465, 678, 928, 965, 313, 276, 792, 587, 647, 855, 424,
        22, 195, 283,  56, 323,  52, 657, 164, 345, 873, 991, 682, 600,
       110, 104, 761, 673, 567, 776, 716, 763, 308, 838,  12, 113, 858,
       704, 649, 366, 447, 474, 183, 258, 629, 910,  98, 700, 425, 681,
       825, 770, 631, 372,  58, 679, 215, 876, 312, 544, 956, 694, 218,
       435, 902, 950, 436, 453, 107, 561, 859, 893, 932, 596, 935, 660,
        29, 555, 459, 418, 278, 163, 384, 856, 812, 417, 479,  62, 375,
       610, 640, 262, 353, 349, 974, 126,  89, 598, 316, 144,  60, 813,
       510, 214, 536, 413, 982, 234, 818, 626, 529, 553, 721, 833, 444,
       677, 411, 732, 971, 577,  28, 490, 589, 394, 955, 794, 272, 802,
       849, 672, 773, 916, 702, 654, 990,  67, 416, 760, 344, 513, 244,
       350, 118, 734, 269, 151, 615, 926, 509, 256, 305, 684, 713, 403,
       193,  68, 933, 979, 701, 167, 843, 488, 223, 210, 680, 102, 759,
       329, 602, 909, 786, 807, 748, 119, 652,  15, 177, 612, 630,  45,
       735, 154, 477, 966, 745, 617, 463, 505, 352, 166, 620,  30, 263,
       997, 367, 605, 483, 170, 226, 504, 401, 393, 753, 731,  16, 396,
       737, 237, 823,  94, 986, 530, 116, 811, 586, 516, 765, 506, 801,
       558, 791, 190, 730, 944, 941, 434,  48, 120, 930, 138, 360, 839,
       376, 527, 581, 472, 507, 480,  76, 171, 289, 875,  17, 257, 865,
       628, 451, 877, 618,   5, 937, 964, 768, 219, 192, 333, 906, 339,
       741, 134, 656,  25, 622, 429, 582, 774, 271,  21, 254, 235, 943,
       822, 899, 887, 535, 471, 169,  66, 805, 550, 799, 815, 712, 661,
       638,  57, 767, 832, 264, 331, 580, 284, 176, 572,  51, 398, 386,
         0, 238,  26, 883, 222, 814, 282, 383, 378, 566, 531, 746, 252,
       427, 357, 988,  31, 454,   7, 862, 725, 948, 809, 181, 565, 992,
       798, 137, 295, 972, 771, 766, 334, 301, 720, 970, 953,  13, 777,
       850, 723, 275, 969, 758, 197, 412, 747, 796, 924, 130, 946, 597,
       846, 665, 879, 787, 981, 358, 819, 291,  33, 161, 231, 663, 519,
        69, 297, 959, 517, 912, 455, 841, 160,  81, 623, 706, 659, 772,
       260, 790, 918,  75, 958, 296, 280, 199, 695, 960, 556,  59, 952,
       788, 750, 139, 727, 217, 646,  90, 889, 639, 515, 448, 650, 395,
       129, 320, 870, 525, 494, 637, 124, 426,  18, 317, 319,  65, 613,
        53, 545, 422, 111, 106, 817, 466, 978, 977, 512, 149, 885, 294,
       152, 327, 726, 204, 724, 370, 925, 668, 445,  27,  10, 627, 279,
       285, 588, 585,  32, 482, 784, 132, 845, 762, 499, 662, 621,  39,
       122, 806, 281, 476, 658, 189,   6, 304, 808, 675, 123, 837, 641,
       188, 267, 168, 268, 311, 324, 998,  42, 147, 328, 619, 895, 540,
        20, 501, 438, 180,  41, 140, 707,  47, 816, 840, 249,  97, 495,
       967, 310, 708, 624, 330, 356, 332, 520, 601, 847, 185,  77, 715,
       939, 196, 584, 203, 382, 576, 644, 542, 789, 470,  64, 156])

**Comment by dylan-campbell** [comment]
Ah I see. I don't remember exactly, but I think they are the indices of the 2D points that you would need to take such that they correspond to the ordered list of 3D points.
